src/Spanner/Database.php

```diff
@@ -0,0 +1,1313 @@
+     * strong consistency.
+     *
+     * Snapshots are executed behind the scenes using a Read-Only Transaction.
+     *
+     * Example:
+     * ```
+     * $snapshot = $database->snapshot();
+     * ```
+     *
+     * ```
+     * // Take a shapshot with a returned timestamp.
+     * $snapshot = $database->snapshot([
+     *     'returnReadTimestamp' => true
+     * ]);
+     *
+     * $timestamp = $snapshot->readTimestamp();
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/spanner/reference/rpc/google.spanner.v1#google.spanner.v1.BeginTransactionRequest BeginTransactionRequest
+     * @see https://cloud.google.com/spanner/docs/transactions Transactions
+     *
+     * @param array $options [optional] {
+     *     Configuration Options
+     *
+     *     See [ReadOnly](https://cloud.google.com/spanner/reference/rpc/google.spanner.v1#google.spanner.v1.TransactionOptions.ReadOnly)
+     *     for detailed description of available options.
+     *
+     *     Please note that only one of `$strong`, `$readTimestamp` or
+     *     `$exactStaleness` may be set in a request.
+     *
+     *     @type bool $returnReadTimestamp If true, the Cloud Spanner-selected
+     *           read timestamp is included in the Transaction message that
+     *           describes the transaction.
+     *     @type bool $strong Read at a timestamp where all previously committed
+     *           transactions are visible.
+     *     @type Timestamp $readTimestamp Executes all reads at the given
+     *           timestamp.
+     *     @type Duration $exactStaleness Represents a number of seconds. Executes
+     *           all reads at a timestamp that is $exactStaleness old.
+     * }
+     * @return Snapshot
+     * @codingStandardsIgnoreEnd
+     */
+    public function snapshot(array $options = [])
+    {
+        // These are only available in single-use transactions.
+        if (isset($options['maxStaleness']) || isset($options['minReadTimestamp'])) {
+            throw new \BadMethodCallException(
+                'maxStaleness and minReadTimestamp are only available in single-use transactions.'
+            );
+        }
+
+        $transactionOptions = $this->configureSnapshotOptions($options);
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READ);
+
+        try {
+            return $this->operation->snapshot($session, $transactionOptions);
+        } finally {
+            $session->setExpiration();
+        }
+    }
+
+    /**
+     * Execute Read/Write operations inside a Transaction.
+     *
+     * Using this method and providing a callable operation provides certain
+     * benefits including automatic retry when a transaction fails. In case of a
+     * failure, all transaction operations, including reads, are re-applied in a
+     * new transaction.
+     *
+     * If a transaction exceeds the maximum number of retries,
```

Comment:
The convention we have everywhere else in the veneer libraries is a bounded number of retries. Since the introduction of the GAPIC layer, we have discussed porting the veneer over to use deadlines instead, and plan to do so eventually.